### PR TITLE
storage: perform csr operations off the timely thread

### PR DESCRIPTION
**Pull request by guswynn**
Part of: https://github.com/MaterializeInc/materialize/issues/25707

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by benesch — APPROVED** [review]
One down! Thank you, Gus!

File: src/storage/src/decode/mod.rs [hunk 1]
```diff
@@ -315,7 +316,15 @@ async fn get_decoder(
...
                 None => None,
-                Some(csr_connection) => Some(csr_connection.connect(storage_configuration).await?),
+                Some(csr_connection) => {
+                    let debug_name = debug_name.to_string();
+                    let storage_configuration = storage_configuration.clone();
+                    Some(
+                        async move { csr_connection.connect(&storage_configuration).await }
+                            .run_in_task(move || format!("connect_csr:{}", debug_name))
+                            .await?,
+                    )
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/storage/src/decode/mod.rs:326]: Yeah I can break it out

File: src/interchange/src/avro/schema.rs [hunk 2]
```diff
@@ -410,21 +412,31 @@ impl SchemaCache {
...
                 let response = Retry::default()
-                    .max_duration(Duration::from_secs(30))
-                    .retry_async(|state| async move {
-                        let res = ccsr_client.get_schema_by_id(id).await;
-                        match res {
-                            Err(e) => {
-                                if let Some(timeout) = state.next_backoff {
-                                    warn!("transient failure fetching schema id {}: {:?}, retrying in {:?}", id, e, timeout);
+                    // The timeout of ccsr requests is 60s.
+                    .max_duration(Duration::from_secs(120))
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/interchange/src/avro/schema.rs:418]: yeah I can do that

File: src/ore/src/future.rs [hunk 3]
```diff
@@ -81,6 +92,16 @@ where
         }
     }
 
+    async fn run_in_task<Name, NameClosure>(self, nc: NameClosure) -> T::Output
```
Comment (moulimukherjee) [inline src/ore/src/future.rs:95]: nice! really handy

**Pull request merged**